grudge/trace_pair.py

```diff
@@ -276,57 +284,109 @@ class _RankBoundaryCommunication:
+        # re-containerize the exchanged result
+        self.container_data = container_data
+        _local_data_numpy = np.concatenate(flattened_data)
 
-        local_data = self.array_context.to_numpy(flatten(self.local_dof_array))
         comm = self.dcoll.mpi_communicator
 
-        self.send_req = comm.Isend(local_data, remote_rank, tag=self.tag)
-        self.remote_data_host = np.empty_like(local_data)
-        self.recv_req = comm.Irecv(self.remote_data_host, remote_rank, self.tag)
+        self.send_req = comm.Isend(_local_data_numpy, remote_rank, tag=self.tag)
+        self.remote_data_host_numpy = np.empty_like(_local_data_numpy)
+        self.recv_req = comm.Irecv(
+            self.remote_data_host_numpy, remote_rank, self.tag
+        )
 
     def finish(self):
         self.recv_req.Wait()
 
+        # Re-containerize the recieved data
         actx = self.array_context
-        remote_dof_array = unflatten(
-            self.array_context, self.bdry_discr,
-            actx.from_numpy(self.remote_data_host)
+        discr = self.bdry_discr
+        remote_data_numpy = self.remote_data_host_numpy
+        original_container_data = self.container_data
+        # Convert flattend (concatenated) array data into a list of arrays/obj arrays
+        # denoting the components of the array container
+        remote_container_data = {}
+
+        # This is why we want the original container dict to be ordered
+        # (looping over standard dicts is not parallel safe since keys/items are
+        # not ordered).
+        # NOTE: This assumes all sub arrays are of the same shape!
+        offset = 0
+        for key in original_container_data:
+            _value = original_container_data[key]
+            # Special case when container component is expected to be an object array
+            if isinstance(_value, np.ndarray) and _value.dtype == "O":
+                remote_comps = []
+                for _subary_idx in range(len(_value)):
+                    _subary = _value[_subary_idx]
+                    _subary_ndofs, = _subary.shape
+                    remote_comps.append(
+                        unflatten_from_numpy(
+                            actx, discr,
+                            remote_data_numpy[offset:offset+_subary_ndofs]
+                        )
+                    )
+                    offset += _subary_ndofs
+                remote_data = make_obj_array(remote_comps)
+            else:
+                _ndofs, = _value.shape
+                remote_data = unflatten_from_numpy(
+                    actx, discr, remote_data_numpy[offset:offset+_ndofs]
+                )
+                offset += _ndofs
+            # Store the unflattened, de-numpy-fied remote data
+            remote_container_data[key] = remote_data
+
+        # Re-containerize the remote data
+        remote_array_ctr = deserialize_container(
+            self.local_array_ctr,
+            remote_container_data.items()
         )
```

Comment:
@inducer I'm very unhappy with how this turned out, but it seems to work. However, I think it is too inefficient (performance seems to get worse the more ranks are communicating). The challenge here was trying to "unconvert" an array container whose components have been flattened and concatenated into a single numpy array to be communicated (see https://github.com/inducer/grudge/pull/154/files#diff-5fa6e6b713d37028c16f1bbc5a6d0b4547ecba0f86c8a52eb30e1ba1dc2614e4R322-R330).

This needs to be revisited before I am comfortable merging this. I would appreciate some insight on this.

Reply:
Would it be clearer make something like [unflatten_like](https://github.com/inducer/meshmode/blob/04bcfba6a94091c90d488b628a516d0316a1e5f3/meshmode/dof_array.py#L557-L612) in meshmode for this? That basically works by passing in a prototype array container (i.e. just keep a reference to the original container you flattened) and reconstructs the flat version from that, instead of keeping track of all the offsets.

It could be worth extending `flatten` (or just `flatten_to_numpy`?) with a flag to force it to flatten the complete array container, something like
```
result = []
rec_map_reduce_array_container(
	reduce_func=result.append, 
	map_func=lambda x: x, 
	flatten_to_numpy(ary))
np.concatenate(result)
```

Just some thoughts, not quite sure what the best way to handle this would be..

Reply:
@alexfikl Having something like `unflatten_like` would be very helpful. Unfortunately, I am not quite as familiar with the array traversal infrastructure in place as well as yourself or @inducer --- so I am not sure if something like this already exists within the current functionality.

Really at the end of the day, I have two problems here:

1. What is a good numpy representation of an arbitrary array container?
2. Define an invertible mapping from arbitrary array container to numpy representation.

My solution to 1. is to simply flatten the entire data structure. 2. is trickier because there "inverse" transformation is not so straightforward (at least to me). It sounds like the inverse mapping I want is very similar to what `unflatten_like` is doing. However, I don't recall it working right when I attempted to use it. I'll need to revisit it...

Reply:
Hence my reason for asking for help! :smile:

Reply:
Yeah, the current `flatten`/`unflatten` functions don't do what you need here. They traverse the array container tree and just work on the `DOFArray`s, leaving the rest of the tree as is.

Reply:
Tried cleaning up in https://github.com/inducer/grudge/pull/154/commits/e37462caddf99d48c9476c14f976da9b670afec4